### ISSUE TITLE
Possible regression related to circular types using mapped types and intersections in TypeScript 3.9

### ISSUE THREAD

**Issue by butchler** [microsoft/TypeScript#39147]
**TypeScript Version:**  3.9.2

**Search Terms:**
"Type alias circularly references itself"
"recursive type"
"circular type"
"TypeScript 3.9"
"regression"
"intersection"

**Code**

```ts
type Fails<T> = T extends Record<string, unknown> ? Readonly<{
  [P in keyof T]: Fails<T[P]>;
} & {
  extra: unknown;
}> : T;
```

**Expected behavior:**
There is no type error and the circular type is usable.

**Actual behavior:**
Fails with `Type alias 'Fails' circularly references itself.` in TypeScript 3.9 and Nightly (but does not in 3.8 and earlier)

**Playground Link:** 
Playground also contains several alternate versions that do not produce a type error:
https://www.typescriptlang.org/play/#code/PTAEHUHsCcGsGdQEsB2oDMA6AHKARgK4AuoAZgIZIA2iqGmAnAFBECeADgKagBilNAHgAqAPlABeUENCcAHkU4oAJogBKnAMYwlA+EWioA5gBpQBFLBSQA7ijEB+UOvJLIKKqwEBvJqFABtAAVkNFhOVkhSKQBdAC5efnhhIOiRAG4mAF9QADJQHz85fXJ480sbFAzMsXihDKYQUAA5SBICeCNQcjRUBWh4TSIkN1BrGAQWDm4oOHgARmExSWkixRUnTW1dfSNTMqtbBw2XNw9vXwDgujCIqOkAH1AAciLocie40GDV5UQX+TeT1Ajn2FVA8RmCAWQhS6SyNSk9UaLTaHRQhi6oAAtuR2FwlKA2FxQAAKJCYTiYY6udysACUoBGRAAFtxepx+oNhmgxrNJsTIfAAEyLCRSGTyNZqTbQHR6AzovYWA52YH5C5BEKgG6RGIQ8bC5KBVJVXLqwoAkpmZUVU21JFgAAiSHlSEIQ3RhNZ2Nx+MJUy6GmgkHgiBZbJQfQGGiGI15EyJ0wN6FFywlCl+Gy0su2CpM1vKhzVzhpZwKly1Orun0FKZhxrh1TNJdOnnLrytoNsVQRdSYTC0KD0hM4ejm8T41CSXi68RneHiKAIWLwHNA2SbYpnVvn8Tmpg78SF64Plvi6HXGUHw4UeiF+tmC23c-wi+Xq+g6-XS3ys9-C9Afd0zeI8T2Aq0L0yK83BvUciHQB8EBFZ9-zfFc1w3H8UN3QDT2KUDMjwkCMEvAcYJIW8iAAFkQ+AU2w19QCXdDP0wrc-xwoDD1AY9CPA89LyAA

**Related Issues:** 

* https://github.com/piotrwitek/utility-types/issues/144

**Comment by RyanCavanaugh** [comment]
The type alias is circular; < 3.8 had a bug that the circularity was not properly observed during type alias resolution

**Comment by butchler** [comment]
Thanks, just wanted to confirm that this is the intended behavior.

Why are the alternate versions in the playground not considered circular, though? As far as I can tell, they all represent the same circular type:

```ts
// Not using an intersection works
type Works1<T> = T extends Record<string, unknown> ? Readonly<{
  [P in keyof T | 'extra']: P extends 'extra' ? unknown : Works1<T[P]>;
}> : T;

// Not using a mapped type (i.e. Readonly) on the intersection works
type Works2<T> = T extends Record<string, unknown> ? {
  [P in keyof T]: Works2<T[P]>;
} & {
  extra: unknown;
} : T;

// Distributing the mapped type across the intersection works
type Works3<T> = T extends Record<string, unknown> ? Readonly<{
  [P in keyof T]: Works3<T[P]>;
}> & Readonly<{
  extra: unknown;
}> : T;
```

Is it a bug that TypeScript still does not recognize that these are circular too?

**Comment by RyanCavanaugh** [comment]
Certain things can be deferred if there's a different declaration to "pause" the circularity at. I don't want to commit any specific circular behavior to bug/not bug; it depends on a lot of different factors and is sensitive to exactly where TS tries to resolve certain conditionals.

**Comment by butchler** [comment]
Thanks for the clarification :) Closing
